service/graph.go

```diff
@@ -31,21 +32,236 @@ var _ pipelines = (*pipelinesGraph)(nil)
 type pipelinesGraph struct {
 	// All component instances represented as nodes, with directed edges indicating data flow.
 	componentGraph *simple.DirectedGraph
+
+	// Keep track of how nodes relate to pipelines, so we can declare edges in the graph.
+	pipelines map[component.ID]*pipelineNodes
+}
+
+func buildPipelinesGraph(ctx context.Context, set pipelinesSettings) (pipelines, error) {
+	pipelines := &pipelinesGraph{
+		componentGraph: simple.NewDirectedGraph(),
+		pipelines:      make(map[component.ID]*pipelineNodes, len(set.PipelineConfigs)),
+	}
+	for pipelineID := range set.PipelineConfigs {
+		pipelines.pipelines[pipelineID] = &pipelineNodes{
+			receivers: make(map[int64]graph.Node),
+			exporters: make(map[int64]graph.Node),
+		}
+	}
+
+	if err := pipelines.createNodes(set); err != nil {
+		return nil, err
+	}
+	pipelines.createEdges()
+	if err := pipelines.buildNodes(ctx, set); err != nil {
+		return nil, err
+	}
+	return pipelines, nil
+}
+
+// Creates a node for each instance of a component and adds it to the graph
+func (g *pipelinesGraph) createNodes(set pipelinesSettings) error {
+
+	// map[connectorID]pipelineIDs
+	// Keep track of connectors and where they are used.
+	connectorsAsExporter := make(map[component.ID][]component.ID)
+	connectorsAsReceiver := make(map[component.ID][]component.ID)
+
+	for pipelineID, pipelineCfg := range set.PipelineConfigs {
+		for _, recvID := range pipelineCfg.Receivers {
+			if set.Connectors.IsConfigured(recvID) {
+				connectorsAsReceiver[recvID] = append(connectorsAsReceiver[recvID], pipelineID)
+				continue
+			}
+			g.addReceiver(pipelineID, recvID)
+		}
+		for _, procID := range pipelineCfg.Processors {
+			g.addProcessor(pipelineID, procID)
+		}
+		for _, exprID := range pipelineCfg.Exporters {
+			if set.Connectors.IsConfigured(exprID) {
+				connectorsAsExporter[exprID] = append(connectorsAsExporter[exprID], pipelineID)
+				continue
+			}
+			g.addExporter(pipelineID, exprID)
+		}
+	}
+
+	if len(connectorsAsExporter) != len(connectorsAsReceiver) {
+		return fmt.Errorf("each connector must be used as both receiver and exporter")
+	}
+	for connID, exprPipelineIDs := range connectorsAsExporter {
+		rcvrPipelineIDs, ok := connectorsAsReceiver[connID]
+		if !ok {
+			return fmt.Errorf("connector %q must be used as receiver, only found as exporter", connID)
+		}
+		for _, eID := range exprPipelineIDs {
+			for _, rID := range rcvrPipelineIDs {
+				g.addConnector(eID, rID, connID)
+			}
+		}
+	}
+	return nil
+}
+
+func (g *pipelinesGraph) addReceiver(pipelineID, recvID component.ID) {
+	node := newReceiverNode(pipelineID, recvID)
+	if rcvrNode := g.componentGraph.Node(node.ID()); rcvrNode != nil {
+		g.pipelines[pipelineID].addReceiver(rcvrNode)
+		return
+	}
+	g.pipelines[pipelineID].addReceiver(node)
+	g.componentGraph.AddNode(node)
+}
```

Comment:
I would refactor this to be just "createReceiver"
```suggestion
func (g *pipelinesGraph) createReceiver(pipelineID, recvID component.ID) *receiverNode {
	node := newReceiverNode(pipelineID, recvID)
	if rcvrNode := g.componentGraph.Node(node.ID()); rcvrNode != nil {
		return
	}
	g.componentGraph.AddNode(node)
}
```

And on line 70, to separate the creation of the individual nodes with the adding to the pipelines:


```golang
	for pipelineID, pipelineCfg := range set.PipelineConfigs {
		pipe := g.pipelines[pipelineID]
		for _, recvID := range pipelineCfg.Receivers {
			if set.Connectors.IsConfigured(recvID) {
				connectorsAsReceiver[recvID] = append(connectorsAsReceiver[recvID], pipelineID)
				continue
			}
			pipe.addReceiver(g.createReceiver(pipelineID, recvID))
		}
		...
	}
```

Reply:
Just saw your comment. You probably saw by now that I refactored this like you suggested.